compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.kt

```diff
@@ -0,0 +1,230 @@
+import androidx.compose.ui.scene.skia.SwingSkiaLayerComponent
+import androidx.compose.ui.scene.skia.WindowSkiaLayerComponent
+import androidx.compose.ui.window.WindowExceptionHandler
+import androidx.compose.ui.window.density
+import androidx.compose.ui.window.layoutDirectionFor
+import androidx.compose.ui.window.scaledSize
+import java.awt.Component
+import java.awt.Rectangle
+import java.awt.Window
+import java.awt.event.ComponentEvent
+import java.awt.event.ComponentListener
+import java.awt.event.WindowEvent
+import java.awt.event.WindowFocusListener
+import javax.swing.JLayeredPane
+import javax.swing.SwingUtilities
+import kotlin.coroutines.AbstractCoroutineContextElement
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineExceptionHandler
+import org.jetbrains.skiko.MainUIDispatcher
+import org.jetbrains.skiko.SkiaLayerAnalytics
+
+/**
+ * Internal entry point to Compose.
+ *
+ * It binds Skia canvas and [ComposeScene] to [container].
+ * It also configures compose based on [ComposeFeatureFlags].
+ */
+internal class ComposeContainer(
+    val container: JLayeredPane,
+    private val skiaLayerAnalytics: SkiaLayerAnalytics,
+    window: Window? = null,
+
+    private val useSwingGraphics: Boolean = ComposeFeatureFlags.useSwingGraphics,
+) : ComponentListener, WindowFocusListener {
+    val windowContext = PlatformWindowContext()
+    var window: Window? = null
+        private set
+    private var layoutDirection = layoutDirectionFor(window ?: container)
+
+    private val coroutineExceptionHandler = DesktopCoroutineExceptionHandler()
+    private val coroutineContext = MainUIDispatcher + coroutineExceptionHandler
+
+    private val bridge = ComposeBridge(
+        container = container,
+        windowContext = windowContext,
+        exceptionHandler = {
+            exceptionHandler?.onException(it) ?: throw it
+        },
+        coroutineContext = coroutineContext,
+        skiaLayerComponentFactory = ::createSkiaLayerComponent,
+        composeSceneFactory = ::createComposeScene,
+    )
+
+    val contentComponent by bridge::contentComponent
+    val focusManager by bridge::focusManager
+    val accessible by bridge::accessible
+    var rootForTestListener by bridge::rootForTestListener
+    // TODO: Changing fullscreen probably will require recreate our layers
+    //  It will require add this flag as remember parameters in rememberComposeSceneLayer
+    var fullscreen by bridge.skiaLayerComponent::fullscreen
+    var compositionLocalContext by bridge::compositionLocalContext
+    var exceptionHandler: WindowExceptionHandler? = null
+    val windowHandle by bridge.skiaLayerComponent::windowHandle
+    val renderApi by bridge.skiaLayerComponent::renderApi
+    val preferredSize by bridge::preferredSize
+
+    init {
+        setWindow(window)
+    }
+
+    fun dispose() {
+        bridge.dispose()
+    }
+
+    override fun componentResized(e: ComponentEvent?) = onChangeWindowBounds()
```

Comment:
Do we need to subscribe to componentResized/componentMoved, if we already have `setBounds`?

Reply:
Yes. `setBounds` is about changing container, not a window

Reply:
In this case, see my other comment about inderection. We can subscribe to componentResized inside `ComposeWindowPanel`